bsip-0040.md

```diff
@@ -171,9 +180,13 @@ Adding a custom active authority means increased effort for the backend, and wit
 - `install_custom_active_authority`: Normal accounts can only create custom active authoritites with a duration of maximum 1 year, LTM can do any duration. Two options come to mind:
   - A fixed high fee independent of authorities content
   - Tied to the duration and complexity of the custom active authority. Transaction fee is then `fee = flat_fee + basic_fee * duration` where `basic_fee` is calculated according to complexity (e.g. size of authority, number of restrictions and etc.). Fee is capped at 1 year for LTM.
-- `update_custom_active_authority`: Base fee similar to `account_update` plus dynamic one for any duration changes
+- `update_custom_active_authority`: Base fee similar to `account_update` plus dynamic one for any duration changes, no payback if duration in decreased.
 - `delete_custom_active_authority`: Cheap similar to `limit_order_cancel`
 
+This logic forces the user to think about the desired duration and enforces to put it rather too short than too long. 
+If a custom active is expired, or considered to be too short, extending it is easily doable. 
```

Comment:
Ah I see. It was meant like `duration` refers to the time that is still left, not the actual duration (i.e. `date_to - max(now, date_from)`).

In your example, the reference base duration when updating id `2018-08-20 - 2018-09-01`. So updating to `2018-08-15 - 2018-09-15`  means adding 14 days.

Reply:
`now` is chain state which can't be used when calculating fee with current design principle.

Reply:
I see, good point. The fee would then depend on execution time and not creation time.

Let's simply do it that all time period increase counts, whereas decrease is ignored.

In your example:
The change `2018-08-01` to `2018-08-15` decreases the period, ignore.
THe change `2018-09-01` to `2018-09-15` increases the period, add to fee.

Reply:
Unfortunately the value before change is in chain state so can't be used as well.

Reply:
If fee logic can't contain chain state, then we need something like suppose_now, don't we?

Like either that logic with now, or the update must contain old_from and old_to which are used to calculate and must validate to.be equal to on-chain values, otherwise fail. 

Other ideas?

Reply:
Yes, that's the idea I mentioned above.

BTW, I made an implementation about creation fee:

```
billable fields in a custom authority = 
   { time_from, time_to, authority, a list of restrictions }

a restriction = { member, modifier, function, argument }

struct fee_parameters_type
{
   uint64_t basic_fee        ;
   uint32_t price_per_k_unit ;
};

total units of a custom authority =
   (time_to - time_from) in seconds * items in authority * total units of restrictions

units of a restriction = units of argument

total_fee_required = basic_fee + price_per_k_unit * total_units / 1000;
```

Here is a table for calculation units of different argument types:

Type | Unit | Description
--- | --- | ---
public_key_type | 4 | it uses around 32 bytes to store, 4 = 32 / 8
fc::sha256 | 4 | SHA256 hash, used in blind transfers
string | size() / 8 | length of string in 64-bit blocks, round up
list of strings | sum(units(string)) | total units of strings in list
other simple types | 1 | account id, timestamp, integer and etc
list of non-string types | list.size() * units(type) | fixed length
attribute restriction | units(restriction) | calculate recursively

Note:
* list of non-simple types is not allowed in argument.

Reply:
Then the fee of updating should take argument size and authority size into account, but not only duration difference. Ideally, 
```
update_fee = basic_fee + max( 0, total_new_fee - unused_old_fee )
           = basic_fee + max( 0, 
                new_fee_rate * new_duration * new_auth_size * new_argument_size
              - old_fee_rate * max(old_time_to-now, 0) * old_auth_size * old_argument_size )
```
Since all old data is in chain state, to calculate the difference, we need to put some "supposed" data in the operation, to be simpler, we can directly put the supposed delta in the operation and calculate in background.

Reply:
> Note: I changed format of restriction a bit (mentioned in #86 (comment)),

You basically added a modifier field, which is used to extract the implicit type conversions into explicit ones. Which is totally fine and is covered by my note in the BSIP
"This list of restrictions may be redefined to improve clarity, performance or to reduce complexity of implementation while maintaining the intended functionality"

> Since all old data is in chain state, to calculate the difference, we need to put some "supposed" data in the operation, to be simpler, we can directly put the supposed delta in the operation and calculate in background.

What is best here for implementation I leave to you. One question I have: If the user overpays the transaction fee, will he (or can he) get the rest back?

Reply:
> One question I have: If the user overpays the transaction fee, will he (or can he) get the rest back?

In BitShares fees can always be overpaid, and the system never refund overpaid fees except the "refund order creation fee when order is cancelled" feature.

Reply:
BTW I mentioned fee pooling mechanism earlier but Peter didn't like it at this early stage.

Reply:
I saw that, but that takes it to a new level in terms of complexity. I think user can be expected to think before they create one.